### PR TITLE
fix: resolve protocol parameter type errors across era transitions

### LINKED ISSUE #999: Input errors loading protocol parameters
Recent changes caused some input errors when loading the preview network from genesis.

```
{"time":"2025-11-08T15:43:14.561895072-05:00","level":"WARN","msg":"TX a8fa4293645facb2a0332f4dfc442dff3fc9ca021c95ee908df5d9605e3825be failed validation: pparams are not expected type\npparams are not expected type\npparams are not expected type\npparams are not expected type\npparams are not expected type\npparams are not expected type\npparams are not expected type"}
{"time":"2025-11-08T15:43:14.56541456-05:00","level":"WARN","msg":"TX e3ca57e8f323265742a8f4e79ff9af884c9ff8719bd4f7788adaea4c33ba07b6 failed validation: pparams are not expected type\npparams are not expected type\npparams are not expected type\npparams are not expected type\npparams are not expected type\npparams are not expected type\npparams are not expected type"}
{"time":"2025-11-08T15:43:14.567205264-05:00","level":"ERROR","msg":"failed to process block: process transaction certificates: get certificate deposit: pparams are not expected type"}
```

### PR DESCRIPTION
**Pull request by wolf31o2**
Fixes protocol parameter type mismatch errors that occurred during era transitions and database operations.

## Changes
- Add era-safe `GetEraById()` function to prevent crashes from unknown era IDs
- Implement genesis protocol parameter bootstrapping for empty database scenarios  
- Fix database error handling in `GetPParams` with correct error variable usage
- Support complete era transition chain (Byron -> Shelley -> Allegra -> Mary -> Alonzo -> Babbage -> Conway)
- Add comprehensive test coverage for era ID lookups and edge cases

## Validation
- Functional testing: Successfully processed 40,000+ blocks without errors
- All existing tests passing
- Code quality: zero lint issues, proper formatting
- Real-world validation with `dingo load` and `dingo serve`

Closes #999

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved error handling when retrieving protocol parameters and added nil-checks to prevent crashes.
  * Certificate deposit computation now handles era mismatches gracefully and surfaces unexpected errors.
  * Validation now requires both the k-slot window and historical-validation flag.

* **New Features**
  * Era descriptor lookup by identifier.
  * Genesis protocol-parameter bootstrap across eras and era-aware epoch loading.

* **Tests**
  * Extensive unit tests for era lookup and added test-facing helpers for ledger state setup and inspection.

### CODE REVIEW

File: ledger/certs.go [hunk 1]
```diff
@@ -30,12 +30,22 @@ func (ls *LedgerState) processTransactionCertificates(
...
+			certDeposit, err = ls.currentEra.CertDepositFunc(
+				tmpCert,
+				ls.currentPParams,
+			)
+			if err != nil {
+				return fmt.Errorf("get certificate deposit: %w", err)
+			}
+		} else {
+			// For historical blocks, use zero deposit since these were already validated
+			certDeposit = 0
```
Comment (wolf31o2) [inline ledger/certs.go]: I know this isn't actually correct, but it'll work for now.

**Review by verbotenj — APPROVED** [review]

File: ledger/state.go [hunk 2]
```diff
@@ -712,8 +712,9 @@ func (ls *LedgerState) ledgerProcessBlocks() {
 							cutoffSlot = 0
 						}
 
-						// Validate only if blockSlot ≥ cutoffSlot and skip if it fails
-						if blockSlot >= cutoffSlot {
+						// Validate only if ValidateHistorical is enabled and blockSlot ≥ cutoffSlot
+						if ls.config.ValidateHistorical &&
+							blockSlot >= cutoffSlot {
```
Comment (agaffney) [inline ledger/state.go]: Is this going to work to validate near current tip if ValidateHistorical is disabled?

File: ledger/state.go [hunk 3]
```diff
@@ -921,10 +922,150 @@ func (ls *LedgerState) loadPParams() error {
 	if err != nil {
 		return err
 	}
-	ls.currentPParams = pparams
+	// Only overwrite currentPParams if we found something in the database
+	// or if currentPParams is nil
+	if pparams != nil || ls.currentPParams == nil {
+		ls.currentPParams = pparams
+	}
```
Comment (agaffney) [inline ledger/state.go]: I'm not sure we want to guard against `ls.currentPParams` already being populated. We have other cases where a load from the DB will overwrite existing cached data, such as `loadEpochs()`

File: ledger/state.go [hunk 4]
```diff
@@ -921,10 +922,150 @@ func (ls *LedgerState) loadPParams() error {
...
+			return nil, err
+		}
+		return eras.HardForkConway(ls.config.CardanoNodeConfig, babbagePParams)
+	default:
+		return nil, fmt.Errorf(
+			"unsupported era ID %d for genesis parameter loading",
+			ls.currentEra.Id,
+		)
+	}
+}
```
Comment (agaffney) [inline ledger/state.go]: This whole thing seems to be duplicating existing logic, but in a less DRY way

File: ledger/state.go [hunk 5]
```diff
@@ -947,6 +1133,8 @@ func (ls *LedgerState) loadEpochs(txn *database.Txn) error {
 	}
 	startProtoVersion := shelleyGenesis.ProtocolParameters.ProtocolVersion.Major
 	startEra := eras.ProtocolMajorVersionToEra[startProtoVersion]
+	// Initialize current era to Byron when starting from genesis
+	ls.currentEra = eras.Eras[0] // Byron era
```
Comment (agaffney) [inline ledger/state.go]: Does this not already happen elsewhere?

File: ledger/state.go [hunk 6]
```diff
@@ -1074,6 +1262,52 @@ func (ls *LedgerState) GetCurrentPParams() lcommon.ProtocolParameters {
 	return ls.currentPParams
 }
 
+// Test helper methods - these should only be used in tests
```
Comment (agaffney) [inline ledger/state.go]: If these are only used in tests, we probably shouldn't expose them publicly

File: ledger/state.go [hunk 7]
```diff
@@ -1074,6 +1262,52 @@ func (ls *LedgerState) GetCurrentPParams() lcommon.ProtocolParameters {
...
+}
+
+// ProcessTransactionCertificatesPublic is a test helper to expose processTransactionCertificates
+func (ls *LedgerState) ProcessTransactionCertificatesPublic(
+	txn *database.Txn,
+	blockPoint ocommon.Point,
+	certs []ledger.Certificate,
+) error {
+	return ls.processTransactionCertificates(txn, blockPoint, certs)
+}
```
Comment (agaffney) [inline ledger/state.go]: These functions can affect the ledger state outside of processing TXs from blocks, and we should probably not be exposing them outside the package

**Review by agaffney — CHANGES_REQUESTED** [review]

File: ledger/state.go (hunk 7 above)
Comment (wolf31o2) [reply ledger/state.go]: For this, I am going to fix/handle it in the "unified certificate" PR when I inevitably need to rebase it on top of this code.

**Comment by wolf31o2** [comment]
@coderabbitai review

File: ledger/state.go (hunk 5 above)
Comment (wolf31o2) [reply ledger/state.go]: Only if len(epochs) > 0

File: ledger/state.go [hunk 8]
```diff
@@ -701,7 +701,7 @@ func (ls *LedgerState) ledgerProcessBlocks() {
 						break
 					}
 					// Enable validation using the k-slot window from ShelleyGenesis.
-					if !shouldValidate && i == 0 {
+					if i == 0 {
```
Comment (agaffney) [inline ledger/state.go]: We should keep this guard to ensure it only switches validation ON, and that we don't have to do these calculations for each block batch once validation has been enabled
